codegen/smithy-aws-go-codegen/src/main/java/software/amazon/smithy/aws/go/codegen/customization/S3UpdateEndpoint.java

```diff
@@ -0,0 +1,154 @@
+import software.amazon.smithy.go.codegen.integration.ConfigField;
+import software.amazon.smithy.go.codegen.integration.GoIntegration;
+import software.amazon.smithy.go.codegen.integration.MiddlewareRegistrar;
+import software.amazon.smithy.go.codegen.integration.RuntimeClientPlugin;
+import software.amazon.smithy.model.Model;
+import software.amazon.smithy.model.shapes.OperationShape;
+import software.amazon.smithy.model.shapes.ServiceShape;
+import software.amazon.smithy.model.shapes.ShapeId;
+import software.amazon.smithy.model.shapes.StructureShape;
+import software.amazon.smithy.utils.ListUtils;
+
+public class S3UpdateEndpoint implements GoIntegration {
+    private static final String USE_PATH_STYLE_OPTION = "UsePathStyle";
+    private static final String USE_ACCELERATE_OPTION = "UseAccelerate";
+    private static final String UPDATE_ENDPOINT_ADDER = "addUpdateEndpointMiddleware";
+    private static final String UPDATE_ENDPOINT_INTERNAL_ADDER = "UpdateEndpoint";
+
+    /**
+     * Gets the sort order of the customization from -128 to 127, with lowest
+     * executed first.
+     *
+     * @return Returns the sort order, defaults to -40.
+     */
+    @Override
+    public byte getOrder() {
+        return 127;
+    }
+
+    @Override
+    public void writeAdditionalFiles(
+            GoSettings settings,
+            Model model,
+            SymbolProvider symbolProvider,
+            GoDelegator goDelegator
+    ) {
+        if (!isS3Service(model, settings.getService(model))) {
+            return;
+        }
+
+        goDelegator.useShapeWriter(settings.getService(model), this::writeMiddlewareHelper);
+
+        // Generate getter's for an operation input that can be used to satisfy interfaces
+        for (ShapeId shapeId: settings.getService(model).getOperations()) {
+            OperationShape operation = model.expectShape(shapeId, OperationShape.class);
+            StructureShape input = model.expectShape(operation.getInput().get(), StructureShape.class);
+            goDelegator.useShapeWriter(operation, writer -> {
+                writeInputGetter(writer, symbolProvider, input);
+            });
+        }
+    }
+
+    private void writeInputGetter(GoWriter writer, SymbolProvider symbolProvider, StructureShape input) {
+        // generate bucketGetter if input has a member named Bucket
+        if (input.getMember("Bucket").isPresent()) {
+            // generateBucketGetter
+            // TODO: Look for alternatives so that these Getters are NOT exported?
+            Symbol inputSymbol = symbolProvider.toSymbol(input);
+            writer.writeDocs("GetBucket retrieves the Bucket member value if provided");
+            writer.openBlock("func (s $P) GetBucket() (v string) {", "}", inputSymbol, () -> {
+                writer.write("if s.Bucket == nil { return v }");
+                writer.write("return *s.Bucket");
+            });
+            writer.insertTrailingNewline();
+        }
+    }
```

Comment:
So a couple of thoughts here:
* Customizations that require manipulation of the request based on the value of bucket specified in the top level of the shape should be targeted based on members that are of the `com.amazonaws.s3#BucketName` shape type. So we shouldn't be assuming the bucket is always named `Bucket`.
* Instead of generating a `Getter` on the type shape, we instead should generate an exported method in the client package called `getBucketName`. This method definition would be code generated based on the model to correctly retrieve the top-level bucket parameter from the given operation input shape. As an example
```go

// getBucketNameFromInput retrieves the bucket input parameter from the shape. Returns false if the provided input does not contain a bucket parameter.
func getBucketNameFromInput(input interface{}) (*string, bool) {
	switch i := input.(type) {
	case *GetObject:
		return i.Bucket, true
	case *PutObject
		return i.Bucket, true
	default:
		return nil, false
	}
}
```
Then your customization middleware would continue to remain generic, and would use the helper to retrieve the bucketName shape from the input appropriately. This can then be further used for accesspoint ARNs.  Additionally this  same pattern would be used for s3-control to retrieve fields marked as the accountId.